k2/csrc/host/fsa_util.cc

```diff
@@ -251,46 +247,62 @@ void GetArcIndexes2(const Array2<int32_t *, int32_t> &arc_map,
 
 void StringToFsa::GetSizes(Array2Size<int32_t> *fsa_size) {
   K2_CHECK_NE(fsa_size, nullptr);
-  fsa_size->size1 = fsa_size->size2 = 0;
+  fsa_size->size1 = static_cast<int32_t>(arcs_.size());
+  fsa_size->size2 = num_arcs_;
+}
 
+void StringToFsa::ReadArcsFromString() {
   static constexpr const char *kDelim = " \t";
   std::istringstream is(s_);
   std::string line;
   bool finished = false;  // when the final state is read, set it to true.
-  int32_t num_arcs = 0;
+  num_arcs_ = 0;
   while (std::getline(is, line)) {
     std::vector<std::string> splits;
     SplitStringToVector(line, kDelim, &splits);
     if (splits.empty()) continue;
 
     K2_CHECK_EQ(finished, false);
 
-    auto fields = StringVectorToIntVector(splits);
-    auto num_fields = fields.size();
-    if (num_fields == 3u) {
+    auto num_fields = splits.size();
+    if (num_fields == 3u || num_fields == 4u) {
       Arc arc{};
-      arc.src_state = fields[0];
-      arc.dest_state = fields[1];
-      arc.label = fields[2];
+      arc.src_state = StringToInt(splits[0]);
+      arc.dest_state = StringToInt(splits[1]);
+      arc.label = StringToInt(splits[2]);
+      if (num_fields == 4u) {
+        arc.weight = StringToFloat(splits[3]);
+      }
 
       auto new_size = std::max(arc.src_state, arc.dest_state);
       if (new_size >= arcs_.size()) arcs_.resize(new_size + 1);
 
       arcs_[arc.src_state].push_back(arc);
-      ++num_arcs;
-    } else if (num_fields == 1u) {
+      ++num_arcs_;
+    } else if (num_fields == 1u || num_fields == 2u) {
+      // TODO(guoguo): the original code only supports one final state. Is it
+      //               necessary to supports multiple final states?
       finished = true;
-      K2_CHECK_EQ(fields[0] + 1, static_cast<int32_t>(arcs_.size()));
+      K2_CHECK_EQ(StringToInt(splits[0]) + 1,
+                  static_cast<int32_t>(arcs_.size()));
+      if (num_fields == 2u) {
```

Comment:
For Fsa in `k2`, definitely the final states would have no weight.

Reply:
OK, I was trying to make the code smart, and let it convert the final state to an arc with weight if it sees OpenFST style final state. But I guess simple is good, as long as we have the same assumption of the input FST. I deleted this part of the code.